src/lib/hydrate.ts

```diff
@@ -12,79 +12,266 @@
-        const walker = document.createTreeWalker(
-            container, NodeFilter.SHOW_COMMENT, null, false);
-        let node: Comment|null;
-        while ((node = walker.nextNode() as Comment | null) !== null) {
-          if (node.nodeType === Node.COMMENT_NODE) {
-            if (node.textContent!.startsWith('lit-part')) {
-              // This is an NodePart opening marker
-              const partInfo = {
-                startNode: node,
-                endNode: undefined as unknown as Node,
-              };
-              if (partStack.length > 0) {
-                const parentInfo = partStack[partStack.length - 1];
-                if (parentInfo.children === undefined) {
-                  parentInfo.children = [];
+     userOptions?: Partial<RenderOptions>) => {
+      if (parts.has(container)) {
+        throw new Error('container already contains a live render');
+      }
+
+      // TODO: this matches what render() does, but we should probably reuse
+      // code so that we use the same options object
+      const options: RenderOptions = {
+        templateFactory,
+        ...userOptions,
+      };
+
+      // Since render() creates a NodePart to render into, we'll always have
+      // exactly one root part. We need to hold a reference to it so we can set
+      // it in the parts cache.
+      let rootPart: NodePart|undefined = undefined;
+
+      // When we are in-between node part markers, this is the current NodePart.
+      // It's needed to be able to set the NodePart's endNode when we see a
+      // close marker
+      let currentNodePart: NodePart|undefined = undefined;
+
+      // Used to remember parent template state as we recurse into nested
+      // templates
+      const stack: Array<NodePartState> = [];
+
+      const walker = document.createTreeWalker(
+          container, NodeFilter.SHOW_COMMENT, null, false);
+      let node: Comment|null;
+
+      // Walk the DOM looking for part marker comments
+      while ((node = walker.nextNode() as Comment | null) !== null) {
+        const data = node.data;
+
+        if (data.startsWith('lit-part')) {
+          // This is an NodePart opening marker
+          let value: unknown;
+
+          if (stack.length === 0) {
+            // We don't have a parent part, so this must be the root part
+            if (rootPart !== undefined) {
+              throw new Error('there must be only one root part per container');
+            }
+            rootPart = currentNodePart = new NodePart({
+              templateFactory,
+              ...options,
+            });
+            value = rootValue;
+          } else {
+            const state = stack[stack.length - 1];
+
+            if (state.container) {
+              // We have a current template instace, so use its template
+              // processor to create parts
+              currentNodePart =
+                  state.instance.processor.handleTextExpression(options);
```

Comment:
Just noting, the naming of the template processor API is pretty confusing from this point of view but I guess it's too late to change since that's overridable.  A better name might have been `createAttributeParts` and `createNodePart` maybe, and then here, `createNodePart` would read better.

Reply:
Also worth pointing out, the only reason to use the `processor` is to allow customizable `NodePart`/`AttributePart`s, right?  But the current design sorta has knowledge of those baked into this module.

Maybe `hydrate` needs to take a `hydrationProcessor` or something that goes along with any processor overrides and implements the per-part-type value initialization on L243-257?  Maybe premature, because the SSR-side also has baked-in knowledge of what's in the template processor too... Something to think about.

Reply:
Using the template processor should be good for now, as that's the main customization point. The places where we create parts directly correspond to the same in lit-html, and the value priming maybe should be something we eventually bake into the Part API.